### DISCUSSION TITLE
Wildcard for redirect_uri

### DISCUSSION THREAD

**Discussion by virth** [zitadel/zitadel#3161] [Ideas]
To use preview-deployments with auto generated url's we need to configure the `redirect_uri` with a wildcard. 

Like this: `https://*.vercel.app/api/auth/callback/zitadel`.

If there's a workaround to get preview deployments with zitadel working please let me know.

**Comment by livio-a** [comment]
Hey @virth 

At the moment we check strictly by protocol (string equality). But I definitely see the use case.

It could be an option to allow wildcards when `devMode` is set to true. Like we do for `http://` redirect_uris.
![image](https://user-images.githubusercontent.com/9405495/152546538-84c74fce-1a7c-481a-851e-660d55ea42b6.png)

That would keep users from using it too much in production.

**Reply by fforootd** [reply]
I would also vote for using this only in dev mode. It does not seem wise to use this in an production setting 😁

**Reply by milanvanschaik** [reply]
The discussions going on here might be interesting:

https://github.com/supabase/gotrue/pull/334

https://github.com/supabase/supabase/issues/4808

**Reply by fforootd** [reply]
@milanvanschaik I like your comment here https://github.com/supabase/supabase/issues/4808#issuecomment-1013683069

Thats exactly what we do with the `devmode`. It allows you to break the specification and use unsafe ways but it will warn you 😁 

<img width="484" alt="image" src="https://user-images.githubusercontent.com/9879976/152559079-f01aab4b-41a0-4f45-83b6-76afedcff4f1.png">

**Reply by TakumiHaruta** [reply]
@livio-a @fforootd 

We also want to use a wildcard in redirect_url to let Vercel redirect to Zitadel.
Zitadel seems to use Vercel to host the frontend but how do you guys solve this when the feature branch in preview?

**Reply by fforootd** [reply]
We only use vercel ATM for docs and marketing pages. But since we also want to deploy our portal we are working on making this happen.

Recently we merged a change that was needed with our zitadel/oidc lib https://github.com/zitadel/oidc/pull/328 we track the progress of intergration into zitadel here https://github.com/zitadel/zitadel/issues/5110

**Reply by TakumiHaruta** [reply]
Okay thank you so much!

**Comment by Sytten** [comment]
I have a weird usecase where I dont control the return url since each user can deploy the software on their own domain. Usually the better practice is to put a warning on the login page but allow an “open redirect“.

**Reply by fforootd** [reply]
Maybe an option for this would also be the following.

You could create a client for each customer and alter the redirect URI through ZITADELs APIs.

**Reply by Sytten** [reply]
I think for our use case we will build a custom oauth server for this part to authorize devices and own the UI for that (behind the central login). This is similar to Plex where you can login to any random IP your instance happen to run on, but you have a big yellow warning to make sure your trust the server.

**Reply by fforootd** [reply]
So you think like building an `oauth proxy` that fronts your idp and acts as one client towards the idp but wrapps multipe clients on the upstream.

Interesting approach 😁

**Comment by sevensolutions** [comment]
This would be very cool. ZITADEL has first class support for multi-tenant scenarios but when you run tenants on separate sub-domains this feature would make things more simple.
I can think of the following workaround:

Lets say we have:
https://tenant1.mysaas.com
https://tenant2.mysaas.com

You will add the tenant's sub-domain in the state of the original request.
Then, when the authorization request comes back, you will redirect to a common endpoint (redirect_uri).
Eg. https://login.mysaas.com/callback?code=...&state=...
Now you extract the tenant's sub-domain from the state and redirect to the final domain, including all other query params.
Eg. https://tenant2.mysaas.com/callback?code=...&state=...

Maybe this helps someone :)
